huaweicloud/services/gaussdb/data_source_huaweicloud_gaussdb_nosql_flavors.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/helper/hashcode"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/fmtp"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/logp"
+)
+
+func DataSourceGaussDBNoSQLFlavors() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceGaussDBFlavorsRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"engine_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"cassandra", "mongodb", "influxdb", "redis",
+				}, false),
+				Default: "cassandra",
+			},
+			"engine_version": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"vcpus": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"memory": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"availability_zone": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"flavors": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vcpus": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"memory": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"engine_type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"engine_version": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"availability_zones": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// This method is used to filter all available gaussdb flavors by ram, vcpus, engine version and availability zone.
+func filterAvailableFlavors(d *schema.ResourceData, flavorList []flavors.Flavor,
+	az string) ([]map[string]interface{}, []string, error) {
+	var names []string
+	result := make([]map[string]interface{}, 0)
+
+	filter := map[string]interface{}{
+		"EngineVersion": d.Get("engine_version").(string),
+	}
+	if ram, ok := d.GetOk("memory"); ok {
+		filter["Ram"] = strconv.Itoa(ram.(int))
+	}
+	if vcpus, ok := d.GetOk("vcpus"); ok {
+		filter["Vcpus"] = strconv.Itoa(vcpus.(int))
+	}
+
+	filterFlavors, err := utils.FilterSliceWithField(flavorList, filter)
+	if err != nil {
+		return nil, nil, err
+	}
+	logp.Printf("[DEBUG] The filter is %v and the result is %v", filter, filterFlavors)
+
+	for _, item := range filterFlavors {
+		flavor := item.(flavors.Flavor)
+
+		var azList []string
+		for k, v := range flavor.AzStatus {
+			if v == "normal" && (az == "" || (az != "" && az == k)) {
```

Comment:
i see the API response like this:
````
     "az_status": {
        "cn-north-4a": "Normal",
        "cn-north-4a,cn-north-4c,cn-north-4g": "Not supported",
        "cn-north-4a,cn-north-4b,cn-north-4c": "Normal",
        "cn-north-4g": "Not supported",
        "cn-north-4f": "Not supported",
        "cn-north-4e": "Not supported",
        "cn-north-4d": "Not supported",
        "cn-north-4c": "Not supported",
        "cn-north-4b": "Not supported"
       }
````
How to match  `"cn-north-4a,cn-north-4b,cn-north-4c": "Normal"` ?
What should the status of `cn-north-4c` be ?

Reply:
actually is `normal`

Reply:
```
{
  "availability_zone": [
    "cn-north-4b",
    "cn-north-4f",
    "cn-north-4e",
    "cn-north-4d",
    "cn-north-4a",
    "cn-north-4c",
    "cn-north-4g",
    "cn-north-4a,cn-north-4b,cn-north-4c",
    "cn-north-4a,cn-north-4c,cn-north-4g"
  ],
  "az_status": {
    "cn-north-4a": "normal",
    "cn-north-4a,cn-north-4b,cn-north-4c": "normal",
    "cn-north-4a,cn-north-4c,cn-north-4g": "unsupported",
    "cn-north-4b": "unsupported",
    "cn-north-4c": "unsupported",
    "cn-north-4d": "unsupported",
    "cn-north-4e": "unsupported",
    "cn-north-4f": "unsupported",
    "cn-north-4g": "unsupported"
  },
  "engine_name": "cassandra",
  "engine_version": "3.11",
  "ram": "8",
  "spec_code": "geminidb.cassandra.large.4",
  "vcpus": "2"
},
```

Reply:
`cn-north-4a,cn-north-4b,cn-north-4c` means Multi-AZ (three AZs for high availability), while `cn-north-4c` is just one.